exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb

```diff
@@ -0,0 +1,275 @@
+              response.body # Read and discard body
+              SUCCESS
+            when Net::HTTPServiceUnavailable, Net::HTTPTooManyRequests
+              response.body # Read and discard body
+              redo if backoff?(retry_after: response['Retry-After'], retry_count: retry_count += 1)
+              FAILURE
+            when Net::HTTPRequestTimeOut, Net::HTTPGatewayTimeOut, Net::HTTPBadGateway
+              response.body # Read and discard body
+              redo if backoff?(retry_count: retry_count += 1)
+              FAILURE
+            when Net::HTTPBadRequest, Net::HTTPClientError, Net::HTTPServerError
+              # TODO: decode the body as a google.rpc.Status Protobuf-encoded message when https://github.com/open-telemetry/opentelemetry-collector/issues/1357 is fixed.
+              response.body # Read and discard body
+              FAILURE
+            when Net::HTTPRedirection
+              @http.finish
+              handle_redirect(response['location'])
+              redo if backoff?(retry_after: 0, retry_count: retry_count += 1)
+            else
+              @http.finish
+              FAILURE
+            end
+          rescue Net::OpenTimeout, Net::ReadTimeout
+            retry if backoff?(retry_count: retry_count += 1)
+            return FAILURE
+          end
+        end
+
+        def handle_redirect(location)
+          # TODO: figure out destination and reinitialize @http and @path
+        end
+
+        def untraced
+          @tracer.with_span(OpenTelemetry::Trace::Span.new) { yield }
+        end
+
+        def backoff?(retry_after: nil, retry_count:, reason:)
+          return false if retry_count > RETRY_COUNT
+
+          sleep_interval = nil
+          unless retry_after.nil?
+            sleep_interval =
+              begin
+                Integer(retry_after)
+              rescue ArgumentError
+                nil
+              end
+            sleep_interval ||=
+              begin
+                Time.httpdate(retry_after) - Time.now
+              rescue # rubocop:disable Style/RescueStandardError
+                nil
+              end
+            sleep_interval = nil unless sleep_interval&.positive?
+          end
+          sleep_interval ||= rand(2**retry_count)
+
+          sleep(sleep_interval)
+          true
+        end
+
+        def encode(span_data) # rubocop:disable Metrics/MethodLength
+          Opentelemetry::Proto::Collector::Trace::V1::ExportTraceServiceRequest.encode(
+            Opentelemetry::Proto::Collector::Trace::V1::ExportTraceServiceRequest.new(
+              resource_spans: span_data
+                .group_by(&:resource)
+                .map do |resource, span_datas|
+                  Opentelemetry::Proto::Trace::V1::ResourceSpans.new(
+                    resource: Opentelemetry::Proto::Resource::V1::Resource.new(
+                      attributes: resource.label_enumerator.map { |key, value| as_otlp_key_value(key, value) }
```

Comment:
(same comment applies to instrumentation_library mappings)

Can span's have `nil` values for `resource` and `instrumentation_library`? In that case will invoking `resource.label_enumerator.map` and `il.name` / `il.version` cause an exception? And I guess generally what's the expectation for how to handle span's without a resource or instrumentation library?

Reply:
For `resource`, you’d have to explicitly install the tracer provider with a `nil` resource. Both the configurator and the provider have non nil defaults. The tracer constructor always instantiates an InstrumentationLibrary, so that should be non nil. The Span constructor is considered a private API, so a Span should only be created by the Tracer. Technically, SpanData should only be created by Span. I think it is reasonable to assume that the instrumentation library can’t be nil. I’m less sure about the resource. I’ll re-read the spec to see if it always requires a resource - certainly, using the SDK as designed, the resource should be non nil.

Reply:
Did some light testing of this and as usual you are correct, I think this is fine, apologies for the extra work here. I had missed the [default `@resource` being set to `telemetry_sdk`](https://github.com/open-telemetry/opentelemetry-ruby/blob/181165c9761e013241fb9d6eb5f2eddd8473c7d1/sdk/lib/opentelemetry/sdk/configurator.rb#L30) that was added in v0.5.0 (which is great, btw!). I think we're probably fine here and if people are using the tracer in unholy ways/messing with private api's, they are on their own.